### ISSUE TITLE
ReadableStream for `renderMedia`

### ISSUE THREAD

**Issue by Aslemammad** [remotion-dev/remotion#3399]
Streaming the response rather than a buffer only! 

# Feature Request 🛍️

Hey, Thank you for the phenomenal package! I'm Mohammad and we've been using remotion for the new release of Millionjs after using it, I thought of streaming the video so that instead of huge buffers, videos would be sent to the user chunk by chunk so that way the user can receive the video instantly (without waiting for the whole render to be done) and then the server would consume less memory since the whole buffer won't be written in the memory anymore.

## Use Case

This way we'd support more users and the experience of generating videos on-demand would be better.

## Possible Solution

The thing that Remotion is doing is so innovative and magical that I don't know about the general architecture of process but I just think this feature would be so useful! It may not be even possible :)

**Comment by JonnyBurger** [comment]
Thanks, we can consider this.

Somewhere around here we'd need to add this: https://github.com/remotion-dev/remotion/blob/93560022c4d04bf1e693d4955bf27f42a1a9e55e/packages/renderer/src/stitch-frames-to-video.ts#L401

**Comment by Aslemammad** [comment]
That'd be great, scaling with remotion would be really easy.

**Comment by JonnyBurger** [comment]
I tried it out while working on #3445, but there is a problem.

While some containers like `matroska` support streaming the output, a regular MP4 works differently:
1. The video gets encoded and written to disk
2. Only when the video is finished encoding, FFmpeg will write information to the **start of the file** like duration, and information on how to seek. This information is needed so that in a web browser, you can seek to an arbitrary position in the video without having to download the whole video.

Basically, the whole videos needs to be written first, and a second pass is needed to fix the beginning of the file. Therefore this usecase doesn't allow for "streaming", which is the idea that you can already start sending off data while the back-end of the file is still processing.

I also don't want to disable the fast-seeking feature, because it is undesirable most of the time and with the current API, it would cause confusion if this feature is not available if you opt to choose streaming.

**Comment by Aslemammad** [comment]
@JonnyBurger That makes, this came to my mind and I found out that it's really hard to implement.

But another idea came to mind which I guess it's simpler and the name may be renderStream? The idea is to render  frames partially (with renderFrames) and return a stream for each of these (with the control of the user)

```
const { stream } = renderStream({ // or renderMediaStream
  frames: 30 // 30 frames for each stream push
  offset: 60 // start from frame 60 
  ...
})
// now the user also has control of how many streams should be sent to the stream
```

This way, for the user, we can send a result every 30 frames, and we do not wait for the whole result. I tried using renderFrames to get a thumbnail for the videos and it worked really well and fast, so that's why I think it can be used for that API.
I'm not sure the end result would be mp4 (I don't know if it supports streaming), but basically we would make something compatible with hls.js for instance. 

All of that might be wrong, but the whole idea is to use renderFrames to create those streams. Let me know what you think.

**Comment by JonnyBurger** [comment]
@Aslemammad Similar problem, the frames we get from Chrome are actually in the format of Base64-in-JSON, I don't think we can actually stream the single images.

But `onFrameBuffer` lets you get buffers for each image, which is not so bad either, right?
